scvi/hub/hub_model.py

```diff
@@ -0,0 +1,244 @@
+        ):
+            raise ValueError(
+                "Dataset is too large to upload to the Model. \
+                Please refer to scvi-tools tutorials for how to handle this case."
+            )
+        repo_token = Path(repo_token_path).read_text()
+        if repo_create:
+            create_repo(repo_name, token=repo_token)
+        api = HfApi()
+        # upload the model
+        api.upload_file(
+            path_or_fileobj=self._model_path,
+            path_in_repo=self._model_path.split("/")[-1],
+            repo_id=repo_name,
+            token=repo_token,
+        )
+        # upload the data if it exists
+        if os.path.isfile(self._adata_path):
+            api.upload_file(
+                path_or_fileobj=self._adata_path,
+                path_in_repo=self._adata_path.split("/")[-1],
+                repo_id=repo_name,
+                token=repo_token,
+            )
+        # upload the metadata and model card
+        api.upload_file(
+            # path_or_fileobj=json.dump(asdict(self.metadata), io),
+            path_or_fileobj=self.metadata.to_json(indent=4).encode(),
+            path_in_repo=METADATA_FILE_NAME,
+            repo_id=repo_name,
+            token=repo_token,
+        )
+        # upload the model card
+        self.model_card.push_to_hub(repo_name, token=repo_token)
+
+    @classmethod
+    def pull_from_huggingface_hub(cls, repo_name: str):
+        """Placeholder docstring. TODO complete."""
+        cache_dir = snapshot_download(
+            repo_id=repo_name,
+            allow_patterns=["model.pt", "adata.h5ad", METADATA_FILE_NAME],
+        )
+        model_card = ModelCard.load(repo_name)
+        return cls(cache_dir, model_card)
+
+    def __repr__(self):
+        def eval_obj(obj):
+            return "No" if obj is None else "Yes"
+
+        print(
+            "HubModel with:\n"
+            f"local_dir: {self._local_dir}\n"
+            f"model loaded? {eval_obj(self._model)}\n"
+            f"adata loaded? {eval_obj(self._adata)}\n"
+            f"adata_large loaded? {eval_obj(self._adata_large)}\n"
+            f"metadata:\n{self.metadata}\n"
+            f"model_card:"
+        )
+        rich.print(Markdown(self.model_card.content))
+        return ""
+
+    @property
+    def metadata(self) -> HubMetadata:
+        """Placeholder docstring. TODO complete."""
+        return self._metadata
+
+    @property
+    def model_card(self) -> ModelCard:
+        """Placeholder docstring. TODO complete."""
+        return self._model_card
+
+    @property
+    def model(self) -> Type[BaseModelClass]:
+        """Placeholder docstring. TODO complete."""
+        if self._model is None:
+            self.load_model()
+        return self._model
+
+    @property
+    def adata(self) -> AnnData:
+        """Placeholder docstring. TODO complete."""
+        if self._adata is None:
+            self.read_adata()
+        return self._adata
+
+    @property
```

Comment:
functools.cached_property
(https://docs.python.org/3/library/functools.html#functools.cached_property)

might help clean this up

Reply:
Added to https://github.com/scverse/scvi-tools/issues/1817
See https://github.com/scverse/scvi-tools/pull/1812#discussion_r1045147118